tls/data_source_public_key.go

```diff
@@ -0,0 +1,75 @@
+package tls
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func dataSourcePublicKey() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourcePublicKeyRead,
+		Schema: map[string]*schema.Schema{
+			"private_key": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "PEM formatted string to use as the private key",
+			},
+
+			"private_key_path": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "File path of the PEM formatted string to use as the private key",
+				ForceNew:    true,
+			},
+			"public_key_pem": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"public_key_openssh": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourcePublicKeyRead(d *schema.ResourceData, meta interface{}) error {
+	// Read private key
+	bytes := []byte(d.Get("private_key").(string))
```

Comment:
You could also use [`GetOk`](https://github.com/hashicorp/terraform/blob/master/helper/schema/resource_data.go#L91) or [`GetOkExists`](https://github.com/hashicorp/terraform/blob/master/helper/schema/resource_data.go#L119) here, which are specifically the exact ways you check for a value defined in Terraform.